File: apps/mobile/lib/core/measurements/weight_processing_policy.dart
import 'weight_processing_exception.dart';
import 'weight_processing_method.dart';

final class WeightProcessingPolicy {
  factory WeightProcessingPolicy({
    required int decimalPlaces,
    required WeightProcessingMethod method,
  }) {
    _validateDecimalPlaces(decimalPlaces);
    return WeightProcessingPolicy._(decimalPlaces, method);
  }

  factory WeightProcessingPolicy.parse({
    required int decimalPlaces,
    required String method,
  }) {
    _validateDecimalPlaces(decimalPlaces);
    return WeightProcessingPolicy._(
      decimalPlaces,
      WeightProcessingMethod.parseExact(method),
    );
  }

  const WeightProcessingPolicy._(this.decimalPlaces, this.method);

  final int decimalPlaces;
  final WeightProcessingMethod method;

  static void _validateDecimalPlaces(int decimalPlaces) {
    if (decimalPlaces < 1 || decimalPlaces > 3) {
      throw const WeightProcessingException(
        WeightProcessingException.decimalPlacesInvalid,
        'Weight decimal places must be exactly 1, 2, or 3.',
      );
    }
  }
}

File: services/backend/src/TraderPro.Domain/Common/Measurements/WeightProcessingPolicy.cs
namespace TraderPro.Domain.Common.Measurements;

public sealed record WeightProcessingPolicy
{
    private WeightProcessingPolicy(
        int decimalPlaces,
        WeightProcessingMethod method)
    {
        DecimalPlaces = decimalPlaces;
        Method = method;
    }

    public int DecimalPlaces { get; }

    public WeightProcessingMethod Method { get; }

    public static WeightProcessingPolicy Create(
        int decimalPlaces,
        WeightProcessingMethod method)
    {
        ValidateDecimalPlaces(decimalPlaces);

        if (!Enum.IsDefined(method))
        {
            throw new WeightProcessingException(
                WeightProcessingException.ProcessingMethodInvalid,
                "The weight-processing method must be Standard, Floor, or Ceiling.");
        }

        return new WeightProcessingPolicy(decimalPlaces, method);
    }

    public static WeightProcessingPolicy Parse(
        int decimalPlaces,
        string? method)
    {
        ValidateDecimalPlaces(decimalPlaces);

        var parsedMethod = method switch
        {
            nameof(WeightProcessingMethod.Standard) =>
                WeightProcessingMethod.Standard,
            nameof(WeightProcessingMethod.Floor) =>
                WeightProcessingMethod.Floor,
            nameof(WeightProcessingMethod.Ceiling) =>
                WeightProcessingMethod.Ceiling,
            _ => throw new WeightProcessingException(
                WeightProcessingException.ProcessingMethodInvalid,
                "The weight-processing method must be exactly Standard, Floor, or Ceiling."),
        };

        return new WeightProcessingPolicy(decimalPlaces, parsedMethod);
    }

    private static void ValidateDecimalPlaces(int decimalPlaces)
    {
        if (decimalPlaces is < 1 or > 3)
        {
            throw new WeightProcessingException(
                WeightProcessingException.DecimalPlacesInvalid,
                "Weight decimal places must be exactly 1, 2, or 3.");
        }
    }
}
